### PR TITLE
[doc] Improve type system documentation

### PR DESCRIPTION
**Pull request by qiao-bo**

### CODE REVIEW

File: docs/lang/articles/basic/type.md [hunk 1]
```diff
@@ -35,236 +29,24 @@ For example, the two most commonly used types:
...
+| i64  |:heavy_check_mark:|:large_orange_diamond:|:x:               |:large_orange_diamond:|
+| u8   |:heavy_check_mark:|:x:                   |:heavy_check_mark:|:large_orange_diamond:|
+| u16  |:heavy_check_mark:|:x:                   |:heavy_check_mark:|:large_orange_diamond:|
+| u32  |:heavy_check_mark:|:x:                   |:heavy_check_mark:|:heavy_check_mark:    |
+| u64  |:heavy_check_mark:|:x:                   |:x:               |:large_orange_diamond:|
+| f16  |:heavy_check_mark:|:x:                   |:x:               |:heavy_check_mark:    |
+| f32  |:heavy_check_mark:|:heavy_check_mark:    |:heavy_check_mark:|:heavy_check_mark:    |
+| f64  |:heavy_check_mark:|:heavy_check_mark:    |:x:               |:large_orange_diamond:|
+
+(:large_orange_diamond: requires extension)
```
Comment (mzmzm) [inline docs/lang/articles/basic/type.md:49]:
```suggestion
(:large_orange_diamond: means extension required)
```

File: docs/lang/articles/basic/type.md [hunk 2]
```diff
@@ -35,236 +29,24 @@ For example, the two most commonly used types:
...
-result value.
+| i8   |:heavy_check_mark:|:x:                   |:heavy_check_mark:|:large_orange_diamond:|
+| i16  |:heavy_check_mark:|:x:                   |:heavy_check_mark:|:large_orange_diamond:|
+| i32  |:heavy_check_mark:|:heavy_check_mark:    |:heavy_check_mark:|:heavy_check_mark:    |
+| i64  |:heavy_check_mark:|:large_orange_diamond:|:x:               |:large_orange_diamond:|
+| u8   |:heavy_check_mark:|:x:                   |:heavy_check_mark:|:large_orange_diamond:|
+| u16  |:heavy_check_mark:|:x:                   |:heavy_check_mark:|:large_orange_diamond:|
+| u32  |:heavy_check_mark:|:x:                   |:heavy_check_mark:|:heavy_check_mark:    |
+| u64  |:heavy_check_mark:|:x:                   |:x:               |:large_orange_diamond:|
+| f16  |:heavy_check_mark:|:x:                   |:x:               |:heavy_check_mark:    |
```
Comment (mzmzm) [inline docs/lang/articles/basic/type.md:45]:
```suggestion
```
It looks like we don't have this type

File: docs/lang/articles/basic/type.md [hunk 3]
```diff
@@ -302,16 +84,30 @@ def func(a: ti.f32) -> ti.i64:
...
+
+When using Taichi, you may encounter situations when certain type of data needs
+to be assigned to another type. In this case, type casts are needed. Here, we
+list four use cases commonly seen during Taichi programming, namely implicit
+casts, explicit casts, casting vectors and matrices, and bit-casts.
 
 :::caution
-The type of a variable is **determined on its initialization**.
+In Taichi, the type of a variable is **determined on its initialization**.
+All data types are static in the **Taichi scope**.
```
Comment (mzmzm) [inline docs/lang/articles/basic/type.md:106]:
```suggestion
All data types are static in the **Taichi scope**. (A static type of variable means the type is known at compile time.  This lets the compiler decide whether a given variable can perform the actions requested from it or not. For more information, see [the difference between statically typed and dynamically typed languages](https://stackoverflow.com/questions/1517582/what-is-the-difference-between-statically-typed-and-dynamically-typed-languages)
```

Comment (qiao-bo) [reply docs/lang/articles/basic/type.md:106]: @mzmzm thanks, can you adapt the sentence a bit?
1: (Static here means...... Based on this information, the compiler can decide whether the performed action is legal or not.). 
2: please don't use stackoverflow as an official reference in our document. Wikipedia is fine. cppreference.com is fine. but not random post. Can you try to find another good reference to be linked?

Comment (mzmzm) [reply docs/lang/articles/basic/type.md:106]: Thanks. I update my suggestion.

File: docs/lang/articles/basic/type.md [hunk 4]
```diff
@@ -302,16 +84,30 @@ def func(a: ti.f32) -> ti.i64:
     ...
 ```
 
-### Type casts
+### Type promotion
 
-All data types are static in the **Taichi scope**. Therefore, casts are needed when you want to assign a certain type of data to another one.
+When Taichi performs binary operations on different types, the result is a
+promoted type. This is known as type promotion. Following the C programming
```
Comment (Vissidarte-Herman) [inline docs/lang/articles/basic/type.md:90]: This didn't explain what a promoted type is. Please could you provide more details?

File: docs/lang/articles/basic/type.md [hunk 5]
```diff
@@ -302,16 +84,30 @@ def func(a: ti.f32) -> ti.i64:
...
 :::caution
-The type of a variable is **determined on its initialization**.
+In Taichi, the type of a variable is **determined on its initialization**.
+All data types are static in the **Taichi scope**.
 :::
 
+#### Implicit casts
+
 When a _low-precision_ variable is assigned to a _high-precision_
 variable, it will be implicitly promoted to the _high-precision_ type
```
Comment (Vissidarte-Herman) [inline docs/lang/articles/basic/type.md:112]: Present tense: 

, it is implicitly promoted to the _high-precision_ type without any warning.

File: docs/lang/articles/basic/type.md [hunk 6]
```diff
@@ -302,16 +84,30 @@ def func(a: ti.f32) -> ti.i64:
...
+list four use cases commonly seen during Taichi programming, namely implicit
+casts, explicit casts, casting vectors and matrices, and bit-casts.
 
 :::caution
-The type of a variable is **determined on its initialization**.
+In Taichi, the type of a variable is **determined on its initialization**.
+All data types are static in the **Taichi scope**.
 :::
 
+#### Implicit casts
```
Comment (Vissidarte-Herman) [inline docs/lang/articles/basic/type.md:109]: Implicit cast? 
Singular form or plural form?

File: docs/lang/articles/basic/type.md [hunk 7]
```diff
@@ -323,6 +119,9 @@ def foo():
     a = 1
     print(a)  # 1.0
 ```
+In this example, variable `a` is initialized with `float` type. On the next line, the assign statement will cast `1` from `int` type to `float` type implicitly.
+
```
Comment (Vissidarte-Herman) [inline docs/lang/articles/basic/type.md:123]: In this example, variable `a` is initialized with a `float` type. On the next line, the assignment statement casts `1` from `int` to `float` implicitly.

File: docs/lang/articles/basic/type.md [hunk 8]
```diff
@@ -335,6 +134,8 @@ def foo():
     a = 3.14
     print(a)  # 3
 ```
+In this example, variable `a` is initialized with `int` type. On the next line, the assign statement will cast `3.14` from `float` type to `int` type implicitly.
```
Comment (Vissidarte-Herman) [inline docs/lang/articles/basic/type.md:137]:
```suggestion
In this example, variable `a` is initialized with an `int` type. On the next line, the assignment statement casts `3.14` from `float` to `int` implicitly.
```

File: docs/lang/articles/basic/type.md [hunk 9]
```diff
@@ -402,6 +203,7 @@ my_vec3f = ti.types.vector(3, float)
 my_mat2f = ti.types.matrix(2, 2, float)
 my_ray3f = ti.types.struct(ro=my_vec3f, rd=my_vec3f, l=ti.f32)
 ```
+In this example, we define four compound types which can be used to create fields and create local variables. We can learn how to use it below.
```
Comment (Vissidarte-Herman) [inline docs/lang/articles/basic/type.md:206]:
```suggestion
In this example, we define four compound types for creating fields and local variables. 
```

File: docs/lang/articles/basic/type.md [hunk 10]
```diff
@@ -417,6 +219,7 @@ vec1 = ti.Vector.field(2, dtype=ti.i32, shape=(128, 128, 128))
 mat2 = ti.Matrix.field(2, 2, dtype=ti.i32, shape=(24, 32))
 ray3 = ti.Struct.field({'ro': my_vec3f, 'rd': my_vec3f, 'l': ti.f32}, shape=(1024, 768))
 ```
+In this example, we define three fields which have different types and shapes by two ways. They are exactly the same.
```
Comment (Vissidarte-Herman) [inline docs/lang/articles/basic/type.md:222]:
```suggestion
In this example, we define three fields in two different ways but of exactly the same effect.
```

**Review by Vissidarte-Herman — CHANGES_REQUESTED** [review]

File: docs/lang/articles/basic/type.md [hunk 11]
```diff
@@ -35,236 +29,24 @@ For example, the two most commonly used types:
...
+| i64  |:heavy_check_mark:|:large_orange_diamond:|:x:               |:large_orange_diamond:|
+| u8   |:heavy_check_mark:|:x:                   |:heavy_check_mark:|:large_orange_diamond:|
+| u16  |:heavy_check_mark:|:x:                   |:heavy_check_mark:|:large_orange_diamond:|
+| u32  |:heavy_check_mark:|:x:                   |:heavy_check_mark:|:heavy_check_mark:    |
+| u64  |:heavy_check_mark:|:x:                   |:x:               |:large_orange_diamond:|
+| f16  |:heavy_check_mark:|:x:                   |:x:               |:heavy_check_mark:    |
+| f32  |:heavy_check_mark:|:heavy_check_mark:    |:heavy_check_mark:|:heavy_check_mark:    |
+| f64  |:heavy_check_mark:|:heavy_check_mark:    |:x:               |:large_orange_diamond:|
+
+(:large_orange_diamond: means extension required)
```
[the remark this answers is not in the document]
Comment (qiao-bo) [reply docs/lang/articles/basic/type.md:49]: @mzmzm could you please fix this?

File: docs/lang/articles/basic/type.md [hunk 12]
```diff
@@ -18,15 +18,9 @@ The _category_ can be one of:
...
-
-- `8`
-- `16`
-- `32`
-- `64`
-
+The _digital number_ can be either `8`, `16`, `32`, or `64`.
 It represents how many **bits** are used in storing the data. The larger
-the bit number, the higher the precision is.
+the bit number is, the higher the precision is.
```
[the remark this answers is not in the document]
Comment (qiao-bo) [reply docs/lang/articles/basic/type.md:23]: Why, can you provide a reference please? AFAIK, the two clauses should match. Either all with `is`, or without. With `is` each is a sentence, so i think the changed is better?

File: docs/lang/articles/basic/type.md (hunk 6 above)
Comment (qiao-bo) [reply docs/lang/articles/basic/type.md:109]: Not sure, i am fine as long as they are consistent per level title :/

File: docs/lang/articles/basic/type.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (mzmzm) [reply docs/lang/articles/basic/type.md]: commit ba1bf56de845638137a8129f6002ffa83621d974

**Review by strongoier — COMMENTED** [review]
Shall we add instructions about using `ti.Vector/ti.Matrix/ti.Struct` to construct variables of corresponding types?

File: docs/lang/articles/basic/type.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (Vissidarte-Herman) [reply docs/lang/articles/basic/type.md:23]: Oops. Missed this one. You are right. Please go ahead with your version.

File: docs/lang/articles/basic/type.md (hunk 3 above)
Comment (strongoier) [reply docs/lang/articles/basic/type.md:106]: Did you forget to apply suggestions here @qiao-bo ?

**Review by strongoier — APPROVED** [review]
Mostly looks good to me. I will submit another PR for `ti.Vector/ti.Matrix/ti.Struct` introductions.

File: docs/lang/articles/basic/type.md [hunk 13]
```diff
@@ -35,236 +29,24 @@ For example, the two most commonly used types:
...
+| i64  |:heavy_check_mark:|:large_orange_diamond:|:x:               |:large_orange_diamond:|
+| u8   |:heavy_check_mark:|:x:                   |:heavy_check_mark:|:large_orange_diamond:|
+| u16  |:heavy_check_mark:|:x:                   |:heavy_check_mark:|:large_orange_diamond:|
+| u32  |:heavy_check_mark:|:x:                   |:heavy_check_mark:|:heavy_check_mark:    |
+| u64  |:heavy_check_mark:|:x:                   |:x:               |:large_orange_diamond:|
+| f16  |:heavy_check_mark:|:x:                   |:x:               |:heavy_check_mark:    |
+| f32  |:heavy_check_mark:|:heavy_check_mark:    |:heavy_check_mark:|:heavy_check_mark:    |
+| f64  |:heavy_check_mark:|:heavy_check_mark:    |:x:               |:large_orange_diamond:|
+
+(:large_orange_diamond: Requiring extension.)
```
[the remark this answers is not in the document]
Comment (qiao-bo) [reply docs/lang/articles/basic/type.md:49]: sure, any suggestions?

File: docs/lang/articles/basic/type.md [hunk 14]
```diff
@@ -302,19 +84,36 @@ def func(a: ti.f32) -> ti.i64:
     ...
 ```
 
-### Type casts
+### Type promotion
+
+When Taichi performs binary operations on different types, the result is a
+promoted type. This is known as [type promotion](https://en.wikipedia.org/wiki/Type_conversion#Type_promotion).
+Following the C programming
```
Comment (k-ye) [inline docs/lang/articles/basic/type.md:91]: nit: Can we add a link to the C spec?

Comment (qiao-bo) [reply docs/lang/articles/basic/type.md:91]: @strongoier ;/

**Pull request merged**